semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/asm/JavaTypeSignature.scala

```diff
@@ -0,0 +1,270 @@
+  *
+  * @see https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.9.1
+  */
+sealed trait JavaTypeSignature extends JavaTypeSignature.Printable
+object JavaTypeSignature {
+
+  /** Parse JVM signature using a custom traverser.
+    *
+    * Example:
+    *
+    * {{{
+    *   parse[ClassSignature]("...", new ClassSignatureVisitor)
+    * }}}
+    */
+  final def parse[T](signature: String, visitor: TypedSignatureVisitor[T]): T = {
+    val signatureReader = new SignatureReader(signature)
+    signatureReader.accept(visitor)
+    visitor.result()
+  }
+
+  /** Helper to print parsed JavaTypeSignature back into original signature */
+  trait Printable {
+    def print(sb: StringBuilder): Unit
+    final def pretty: String = {
+      val sb = new StringBuilder
+      this.print(sb)
+      sb.toString
+    }
+  }
+
+  // NOTE: We use Scala-friendly names for base types instead of the Java primitive keywords.
+  abstract class BaseType(val name: String) extends JavaTypeSignature with Product {
+    final override def print(sb: StringBuilder): Unit =
+      sb.append(this.productPrefix)
+  }
+  object BaseType {
+    case object B extends BaseType("Byte")
+    case object C extends BaseType("Char")
+    case object D extends BaseType("Double")
+    case object F extends BaseType("Float")
+    case object I extends BaseType("Int")
+    case object J extends BaseType("Long")
+    case object S extends BaseType("Short")
+    case object Z extends BaseType("Boolean")
+
+    // JVMS deviation. The void descriptor is not part of BaseType
+    // but having it here simplifies things.
+    case object V extends BaseType("Unit")
+  }
+
+  sealed trait ReferenceTypeSignature extends JavaTypeSignature
+
+  case class ClassTypeSignature(
+      simpleClassTypeSignature: SimpleClassTypeSignature,
+      classTypeSignatureSuffix: List[ClassTypeSignatureSuffix]
+  ) extends ReferenceTypeSignature
+      with ThrowsSignature {
+    override def print(sb: StringBuilder): Unit = {
+      sb.append('L')
+      simpleClassTypeSignature.print(sb)
+      classTypeSignatureSuffix.foreach(_.print(sb))
+      sb.append(';')
+    }
+  }
+  object ClassTypeSignature {
+    def simple(name: String): ClassTypeSignature =
+      ClassTypeSignature(
+        SimpleClassTypeSignature(name, None),
+        Nil
+      )
+  }
+
+  case class TypeVariableSignature(identifier: String)
+      extends ReferenceTypeSignature
+      with ThrowsSignature {
+    override def print(sb: StringBuilder): Unit = {
+      sb.append('T')
+      sb.append(identifier)
+      sb.append(';')
+    }
+  }
+
+  case class ArrayTypeSignature(javaTypeSignature: JavaTypeSignature)
+      extends ReferenceTypeSignature {
+    override def print(sb: StringBuilder): Unit = {
+      sb.append("[")
+      javaTypeSignature.print(sb)
+    }
+  }
+
+  // JVMS deviation. There is no PackageSpecifier, we encode the path in the
+  // identifier instead because that's what we get from ASM.
```

Comment:
Also, how hard would it be to support JVMS fully here? Is this a question of an `indexOf` and `substring`, or the reasons are more fundamental? Not suggesting you do that in the PR, just trying to better understand the code.